src/couch_stats/src/csrt_util.erl

```diff
@@ -0,0 +1,459 @@
+    T;
+maybe_add_delta_int(T, Delta) when is_map(Delta) ->
+    maybe_add_delta_int(T, {delta, Delta});
+maybe_add_delta_int(T, {delta, _} = Delta) ->
+    add_delta(T, Delta).
+
+-spec make_delta(PidRef :: maybe_pid_ref()) -> maybe_delta().
+make_delta(undefined) ->
+    undefined;
+make_delta(PidRef) ->
+    TA = get_delta_a(),
+    TB = csrt_server:get_resource(PidRef),
+    Delta = rctx_delta(TA, TB),
+    set_delta_a(TB),
+    Delta.
+
+-spec rctx_delta(TA :: Rctx, TB :: Rctx) -> map().
+rctx_delta(#rctx{} = TA, #rctx{} = TB) ->
+    Delta = #{
+        docs_read => TB#rctx.docs_read - TA#rctx.docs_read,
+        docs_written => TB#rctx.docs_written - TA#rctx.docs_written,
+        js_filter => TB#rctx.js_filter - TA#rctx.js_filter,
+        js_filtered_docs => TB#rctx.js_filtered_docs - TA#rctx.js_filtered_docs,
+        rows_read => TB#rctx.rows_read - TA#rctx.rows_read,
+        changes_returned => TB#rctx.changes_returned - TA#rctx.changes_returned,
+        changes_processed => TB#rctx.changes_processed - TA#rctx.changes_processed,
+        get_kp_node => TB#rctx.get_kp_node - TA#rctx.get_kp_node,
+        get_kv_node => TB#rctx.get_kv_node - TA#rctx.get_kv_node,
+        db_open => TB#rctx.db_open - TA#rctx.db_open,
+        ioq_calls => TB#rctx.ioq_calls - TA#rctx.ioq_calls,
+        dt => make_dt(TA#rctx.updated_at, TB#rctx.updated_at)
+    },
+    %% TODO: reevaluate this decision
+    %% Only return non zero (and also positive) delta fields
+    %% NOTE: this can result in Delta's of the form #{dt => 1}
+    maps:filter(fun(_K, V) -> V > 0 end, Delta);
+rctx_delta(_, _) ->
+    undefined.
+
+-spec get_delta_a() -> maybe_rctx().
+get_delta_a() ->
+    erlang:get(?DELTA_TA).
+
+-spec get_delta_zero() -> maybe_rctx().
+get_delta_zero() ->
+    erlang:get(?DELTA_TZ).
+
+-spec set_delta_a(TA :: rctx()) -> maybe_rctx().
+set_delta_a(TA) ->
+    erlang:put(?DELTA_TA, TA).
+
+-spec set_delta_zero(TZ :: rctx()) -> maybe_rctx().
+set_delta_zero(TZ) ->
+    erlang:put(?DELTA_TZ, TZ).
+
+-spec get_pid_ref() -> maybe_pid_ref().
+get_pid_ref() ->
+    get(?PID_REF).
+
+-spec get_pid_ref(Rctx :: rctx()) -> pid_ref().
+get_pid_ref(#rctx{pid_ref = PidRef}) ->
+    PidRef;
+get_pid_ref(R) ->
+    throw({unexpected, R}).
+
+-spec set_pid_ref(PidRef :: pid_ref()) -> pid_ref().
+set_pid_ref(PidRef) ->
+    erlang:put(?PID_REF, PidRef),
+    PidRef.
+
+%% @equiv set_fabric_init_p(Func, Enabled, true).
+-spec set_fabric_init_p(Func :: atom(), Enabled :: boolean()) -> ok.
+set_fabric_init_p(Func, Enabled) ->
+    set_fabric_init_p(Func, Enabled, true).
+
+%% Expose Persist for use in test cases outside this module
+-spec set_fabric_init_p(Func, Enabled, Persist) -> ok when
+    Func :: atom(), Enabled :: boolean(), Persist :: boolean().
+set_fabric_init_p(Func, Enabled, Persist) ->
+    Key = fabric_conf_key(Func),
+    ok = config:set_boolean(?CSRT_INIT_P, Key, Enabled, Persist).
+
+-spec fabric_conf_key(Key :: atom()) -> string().
```

Comment:
The `fabric_conf_key` is going to be on the hot path. Maybe we should optimize it. Consider adding `-compile({inline, [{fabric_conf_key, 1}]}).`